### ISSUE TITLE
project architecture

### ISSUE THREAD

**Issue by Tovli** [joeaudette/cloudscribe#20]
Hi,
I want to contribute to your project but there are some questions I have regarding your architecture choices:
1) Why should you support different relational DBs? You're basically just duplicating code. It might be more reasonable to build a replaceable data-layer so anyone can just implement his favorite DB data-layer. I would like to see different DB solutions (i.e. NoSQL) rather then different flavors of the same type. 
2) You use a lot of repository classes, Wouldn't it be better to create a IRepository<T> and implement a general repository pattern for all the entities?
3) Why Are you using MSDI and not autofac or any other stable library? Although this project relies on beta product I believe that autofac might be a better choice (of course you can always move to MSDI with some effort in the future)
4) It seems that you rely on many files in your software. The settings are from appsettings.json (except for siteSettings class), the localization resources are from the resx files. and also the menu comes from a json/xml file. It will make you troubles when you would like to create a CMS for the platform.
5) On your core.models project you've created mvc DataAnnotations attributes. From my prespective the core project should consist on none web related object. Think about the future where you would like to create backend processes. 
6) why did you create your own logging pipeline. It seems that log4net/serilog should be a better approach. BTW - you can consider building a similar mechanics like the serilog sinks to use for your data layer

It seems like I'm criticizing but I really appreciate your effort.

Thanks

**Comment by mikeandersun** [comment]
I agree with this:
>
5) On your core.models project you've created mvc DataAnnotations attributes. From my prespective the core project should consist on none web related object. Think about the future where you would like to create backend processes.

We can have POCO class in core.models project and have DB platform specific POCO in data layer project, and map them. 
Model in core.Models project may by different of models in platform specific data layer projects.

**Comment by joeaudette** [comment]
Please consider that this project is in early stages, I'm still working on the design and much will change. It is always a risk when doing open source work in the public that people will criticize it or not like it because they would do things differently. Maybe I should have kept my work private until it is closer to being ready.

Keep in mind also that while Microsoft may be calling their current release RC1, there is still a lot of things not working and things that are going to change for RC2. Myself I'm thinking of the current release as really beta9, and my work is in even earlier stages.

*  1: I have my reasons. Most of the data access code already existed in the mojoPortal project and I'm re-using it because it is good working code that I have available and because I'm trying to make a possible migration path from mojoPortal to cloudscribe. Those extra db implementations such as pgsql and sqlite are going to come in handy on linux as well. I implemented what I want to implement for my reasons.  Anyone can implement their favorite db data layer if they want to, feel free to implement a NoSql data layer yourself, I don't expect to get around to that in the near future. 

*  2: All of my repositories do implement interfaces. Maybe you would prefer some kind of generic 

        IRepository<TModel> 

type of pattern. I'm not sure that would be of benefit, maybe later if I really think all the repositories are that similar then I would go that direction but I'm not feeling it right now.

*  3: I fully intend to use autofac or at least provide a sample with autofac. I pkan to use the decorator pattern to wrap caching repositories around the implementation repositories and that will not be possible using the default DI so I will need to use autofac to do that. However, I'm building a set of components that will eventually all be published as nuget packages. The DI code goes in the startup of the web app which will be owned by the developer not by me so I don't want to impose a specific DI implementation. I will provide an example using autofac but people will be free to use whatever they want and I may provide examples with a few of the popular DI solutions.

*  4: Localization is not working yet in the Microsoft frameworks so it is not yet implemented in cloudscribe. I do intend to use resx files because I know from my experience with the mojoPportal project that using resx files is an easy way to get other people to make translations by copying and translating the english version of the files. And that also make it easier for people to add custom translations or even just customize the labels.

An xml file is a good solution for navigation of things that map to controller actions such as administrative pages. If you read [the readme of my Navigation project](https://github.com/joeaudette/cloudscribe.Web.Navigation) you will see there I mention plans to be able to nest NavigationTreeBuilders such that additional TreeBuilders can be invoked to dynamically add nodes to the navigation tree for such scenarios as a CMS. I'm not currently building a CMS and I have not got to that yet. The xml file for navigation just builds a foundation it does not limit what we can do to add nodes to the navigation. It also makes it easy for people to customize and add their own navigaiton items.

*  5: Yes I've implemented a few little data annotations in the models project but not using them on the models. They currently are only used on ViewModels to facilitate some validation scenarios. You might be right that those should live in the Core.Web project since they are webby. Possibly they will get moved there, but I'm thinking of breaking up the web project into separate smaller projects. For example I've decided the Setup controller and setup related things should be decoupled from the Core.Web project. One advantage of having those annotations in the model project is that is a project that pretty much any of the other web projects will already need to reference so it is convenient for them to be there. I'm not sure I will even use those annotations in the final results, if I do than maybe I will move them into a separate project but for now I'm not all that worried about it.

*  6: I already had some existing data access code for logging to the database so I implemented a custom logger with no dependencies on other loggers. I'm not imposing on anyone to use my logger, people can decide what to use in the startup of their own projects. Logging is another case of something I'm thinking of decoupling from Core.Web and moving into its own projects.

In summary I do appreciate you taking the time to look at my projects and provide feedback, but please understand that things are still in very early stages.

**Comment by Tovli** [comment]
Hi,
Thank you for the prompt response. I realize that this project is on its early stage.
I do want to contribute, though.
Do you have any advice on where to focus? do you run some kind of task management? so we won't work on the same functionality?

**Comment by joeaudette** [comment]
I just put together quickly a [contributing guide](https://github.com/joeaudette/cloudscribe/blob/master/CONTRIBUTING.md)

There are some pre-requisites indicated there such as a copyright transfer agreement that we would need signed before we can review any significant submissions.

If you can comply with that then we can discuss something of limited scope that you could work on as a first contribution. If you have something specific you are interested in working on let me know, otherwise I can suggest maybe you could work on treebuilder functionality for the [Navigation project](https://github.com/joeaudette/cloudscribe.Web.Navigation)
